// pdf417-barcode/esm2015/lib/pdf417-barcode.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild } from '@angular/core';
import { PDF417 } from './pdf417';
export class Pdf417BarcodeComponent {
    constructor() {
        this.text = '';
        this.opts = {};
        this.pdf417 = PDF417;
    }
    /**
     * @private
     * @return {?}
     */
    update() {
        /** @type {?} */
        let aspectratio = this.opts.aspect || 4;
        /** @type {?} */
        let errorcorrection = this.opts.errlvl || -1;
        /** @type {?} */
        let pixelheight = this.opts.pxh || 1;
        /** @type {?} */
        let pixelwidth = this.opts.pxw || 1
        /* update the barcode class */
        ;
        /* update the barcode class */
        this.pdf417.init(this.text, errorcorrection, aspectratio);
        /** @type {?} */
        const barcode = this.pdf417.getBarcodeArray();
        /* create a new canvas and get drawing context */
        /** @type {?} */
        const newcanvas = document.createElement('canvas');
        newcanvas.width = pixelwidth * barcode['num_cols'];
        newcanvas.height = pixelheight * barcode['num_rows'];
        /** @type {?} */
        const drawing = newcanvas.getContext('2d');
        /* print barcode pixels */
        /** @type {?} */
        var y = 0;
        /* for each row */
        for (var r = 0; r < barcode['num_rows']; ++r) {
            /** @type {?} */
            var x = 0;
            /* for each column */
            for (var c = 0; c < barcode['num_cols']; ++c) {
                if (barcode['bcode'][r][c] == 1 && drawing != null) {
                    drawing.fillRect(x, y, pixelwidth, pixelheight);
                }
                x += pixelwidth;
            }
            y += pixelheight;
        }
        /* replace canvas in container */
        /** @type {?} */
        const el = this.container.nativeElement;
        if (el.firstChild)
            el.removeChild(el.firstChild);
        el.appendChild(newcanvas);
    }
    /* handle init and changes of input data */
    /**
     * @return {?}
     */
    ngAfterViewInit() { this.update(); }
    /**
     * @return {?}
     */
    ngOnChanges() { this.update(); }
}
Pdf417BarcodeComponent.decorators = [
    { type: Component, args: [{
                selector: 'pdf417barcode',
                template: `<div #container></div>`
            }] }
];
Pdf417BarcodeComponent.propDecorators = {
    text: [{ type: Input }],
    opts: [{ type: Input }],
    container: [{ type: ViewChild, args: ['container',] }]
};
if (false) {
    /** @type {?} */
    Pdf417BarcodeComponent.prototype.text;
    /** @type {?} */
    Pdf417BarcodeComponent.prototype.opts;
    /** @type {?} */
    Pdf417BarcodeComponent.prototype.container;
    /**
     * @type {?}
     * @private
     */
    Pdf417BarcodeComponent.prototype.pdf417;
}
