// pdf417-barcode/esm5/lib/pdf417-barcode.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild } from '@angular/core';
import { PDF417 } from './pdf417';
var Pdf417BarcodeComponent = /** @class */ (function () {
    function Pdf417BarcodeComponent() {
        this.text = '';
        this.opts = {};
        this.pdf417 = PDF417;
    }
    /**
     * @private
     * @return {?}
     */
    Pdf417BarcodeComponent.prototype.update = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var aspectratio = this.opts.aspect || 4;
        /** @type {?} */
        var errorcorrection = this.opts.errlvl || -1;
        /** @type {?} */
        var pixelheight = this.opts.pxh || 1;
        /** @type {?} */
        var pixelwidth = this.opts.pxw || 1
        /* update the barcode class */
        ;
        /* update the barcode class */
        this.pdf417.init(this.text, errorcorrection, aspectratio);
        /** @type {?} */
        var barcode = this.pdf417.getBarcodeArray();
        /* create a new canvas and get drawing context */
        /** @type {?} */
        var newcanvas = document.createElement('canvas');
        newcanvas.width = pixelwidth * barcode['num_cols'];
        newcanvas.height = pixelheight * barcode['num_rows'];
        /** @type {?} */
        var drawing = newcanvas.getContext('2d');
        /* print barcode pixels */
        /** @type {?} */
        var y = 0;
        /* for each row */
        for (var r = 0; r < barcode['num_rows']; ++r) {
            /** @type {?} */
            var x = 0;
            /* for each column */
            for (var c = 0; c < barcode['num_cols']; ++c) {
                if (barcode['bcode'][r][c] == 1 && drawing != null) {
                    drawing.fillRect(x, y, pixelwidth, pixelheight);
                }
                x += pixelwidth;
            }
            y += pixelheight;
        }
        /* replace canvas in container */
        /** @type {?} */
        var el = this.container.nativeElement;
        if (el.firstChild)
            el.removeChild(el.firstChild);
        el.appendChild(newcanvas);
    };
    /* handle init and changes of input data */
    /* handle init and changes of input data */
    /**
     * @return {?}
     */
    Pdf417BarcodeComponent.prototype.ngAfterViewInit = /* handle init and changes of input data */
    /**
     * @return {?}
     */
    function () { this.update(); };
    /**
     * @return {?}
     */
    Pdf417BarcodeComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () { this.update(); };
    Pdf417BarcodeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'pdf417barcode',
                    template: "<div #container></div>"
                }] }
    ];
    Pdf417BarcodeComponent.propDecorators = {
        text: [{ type: Input }],
        opts: [{ type: Input }],
        container: [{ type: ViewChild, args: ['container',] }]
    };
    return Pdf417BarcodeComponent;
}());
export { Pdf417BarcodeComponent };
if (false) {
    /** @type {?} */
    Pdf417BarcodeComponent.prototype.text;
    /** @type {?} */
    Pdf417BarcodeComponent.prototype.opts;
    /** @type {?} */
    Pdf417BarcodeComponent.prototype.container;
    /**
     * @type {?}
     * @private
     */
    Pdf417BarcodeComponent.prototype.pdf417;
}
